src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +257,300 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
+			return true;
+		}
+
+		fformat(replayCtx->txnBuffer, "%s\n", line);
+
+		/* Close replaybuffer to mark it as complete */
+		fclose(replayCtx->txnBuffer);
+
+		/* Stream the transaction buffer into the target database */
+		if (!streamBufferedXID(replayCtx))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		replayCtx->isTxnBuffering = false;
+
+		/* Remove the symlink to the replay buffer */
+		(void) unlink_file(context->paths.txnlatestfile);
+
+		/* Early exit if we reached the end position */
+		*stop = context->reachedEndPos;
+
+		replayCtx->txnBuffer = NULL;
+	}
+	else
+	{
+		if (replayCtx->skipTxnBuffering)
+		{
+			return stream_replay_line(ctx, line, stop);
+		}
+
+		if (replayCtx->isTxnBuffering)
+		{
+			/*
+			 * We are in a transaction block, buffer all the messages
+			 * including KEEPALIVE, SWITCHWAL and ENPOS.
+			 */
+			fformat(replayCtx->txnBuffer, "%s\n", line);
+		}
+		else if (metadata.action == STREAM_ACTION_KEEPALIVE ||
+				 metadata.action == STREAM_ACTION_SWITCH ||
+				 metadata.action == STREAM_ACTION_ENDPOS)
+		{
+			/*
+			 * We allow KEEPALIVE, SWITCHWAL and ENPOS messages in a
+			 * non-transactional context. In that case, call stream_replay_line
+			 * directly without buffering the message.
+			 */
+			return stream_replay_line(ctx, line, stop);
+		}
+		else
+		{
+			/*
+			 * When the follow switches from prefetch to replay mode, we
+			 * call stream_transform_stream which might stream the partially
+			 * written transaction created during the prefetch mode.
+			 *
+			 * For example, lets consider the partially written C.sql file,
+			 *
+			 * INSERT INTO A VALUES (2);
+			 * COMMIT -- {"xid": 999, "lsn": "0/1230"};
+			 * BEGIN -- {"xid": 1000, "commitLSN": "0/1234"};
+			 * INSERT INTO C VALUES (1);
+			 * INSERT INTO C VALUES (2);
+			 * KEEPALIVE;
+			 *
+			 * After switching to the replay mode, logical decoding will resume
+			 * from consistent point which again starts with a valid transcation
+			 * block.
+			 *
+			 * Lets assume the following contents streamed from the transform
+			 * to catchup process in UNIX PIPE after switching to replay mode,
+			 *
+			 * BEGIN -- {"xid": 999, "commitLSN": "0/1230"};
+			 * INSERT INTO A VALUES (1);
+			 * INSERT INTO A VALUES (2);
+			 * COMMIT -- {"xid": 999, "lsn": "0/1230"};
+			 * BEGIN -- {"xid": 1000, "commitLSN": "0/1234"};
+			 * INSERT INTO C VALUES (1);
+			 * INSERT INTO C VALUES (2);
+			 * COMMIT -- {"xid": 1000, "lsn": "0/1234"};
+			 *
+			 * Contents of C.sql will be streamed by stream_transform_stream
+			 * and followed by the contents from UNIX PIPE. Since both of the
+			 * contents are streamed one after the other, the second block
+			 * contains the full content of previously written
+			 * transaction. So, we can skip previous transaction buffer and
+			 * start buffering from the current transaction.
+			 */
+			log_debug("Received %s when transaction is not "
+					  "in buffering mode", line);
```

Comment:
Please rewrite the message so that it ends with `": %s", line`. This makes it easier to process the logs later.